core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,307 @@
+    l1_batch_number: L1BatchNumber,
+}
+
+async fn get_miniblock_transition_state(
+    connection: &mut StorageProcessor<'_>,
+    miniblock_number: MiniblockNumber,
+) -> L2BlockEnv {
+    let l2_block_info_key = StorageKey::new(
+        AccountTreeId::new(SYSTEM_CONTEXT_ADDRESS),
+        SYSTEM_CONTEXT_CURRENT_L2_BLOCK_INFO_POSITION,
+    );
+    let l2_block_info = connection
+        .storage_web3_dal()
+        .get_historical_value_unchecked(&l2_block_info_key, miniblock_number + 1)
+        .await
+        .unwrap();
+
+    let (next_miniblock_number, next_miniblock_timestamp) =
+        unpack_block_info(h256_to_u256(l2_block_info));
+
+    let miniblock_hash = connection
+        .blocks_web3_dal()
+        .get_miniblock_hash(miniblock_number)
+        .await
+        .unwrap()
+        .unwrap();
+
+    let virtual_blocks = connection
+        .blocks_dal()
+        .get_virtual_blocks_for_miniblock(&(miniblock_number + 1))
+        .await
+        .unwrap()
+        .unwrap();
+
+    L2BlockEnv {
+        number: next_miniblock_number as u32,
+        timestamp: next_miniblock_timestamp,
+        prev_block_hash: miniblock_hash,
+        max_virtual_blocks_to_create: virtual_blocks,
+    }
+}
+
+fn execute_tx<S: ReadStorage>(tx: &Transaction, vm: &mut VmInstance<S, HistoryEnabled>) {
+    vm.make_snapshot();
+    if vm
+        .inspect_transaction_with_bytecode_compression(vec![], tx.clone(), true)
+        .is_ok()
+    {
+        vm.pop_snapshot_no_rollback();
+        return;
+    }
+    vm.rollback_to_the_latest_snapshot();
+    vm.inspect_transaction_with_bytecode_compression(vec![], tx.clone(), false)
+        .expect("Compression can't fail if we don't apply it");
+}
+
+pub struct BasicWitnessInputProducer {
+    connection_pool: ConnectionPool,
+    validation_computational_gas_limit: u32,
+    l2_chain_id: L2ChainId,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+impl BasicWitnessInputProducer {
+    pub async fn new(
+        connection_pool: ConnectionPool,
+        store_factory: &ObjectStoreFactory,
+        state_keeper_config: &StateKeeperConfig,
+        network_config: &NetworkConfig,
+    ) -> anyhow::Result<Self> {
+        // BEWARE, HERE BE DRAGONS.
+        // This usage here is a race condition waiting to happen.
+        // Context -- The VM run in BasicWitnessInputProducer MUST be identical to run from StateKeeper.
+        // Race condition scenario:
+        // 1. StateKeeper runs with a config A, which has value a for `validation_computational_gas_limit`
+        // 2. StateKeeper runs the batch and saves data to database.
+        // 3. Deployment is made and changes to config B, with value b for `validation_computational_gas_limit`.
+        // 4. BasicWitnessInputProducer loads config B
+        // 5. BasicWitnessInputProducer runs the L1Batch that was saved with config A using config B.
+        // In this scenario, the outputs will be different, which will make Witness Generation crash.
```

Comment:
is there a way to fail even earlier? That is - when we generate the basic witness input. do we have `validation_computational_gas_limit` saved per-block anywhere? Shall we add it to the DB? If we do that, we can read that value instead of relying on the config cc @Deniallugo